### PR TITLE
ENH: add decentralized provenance tracking

### PR DESCRIPTION
**Pull request by ebolyen**
Ready for review.

### CODE REVIEW

File: qiime/core/util.py [hunk 1]
```diff
@@ -27,6 +31,50 @@ def decorator(func):
...
+                attr = attr[:-1]
+            results.append("%d %s" % (value, attr))
+    if results:
+        text = results[-1]
+        if results[:-1]:
+            text = ', and '.join([', '.join(results[:-1]), text])
+        return text
+    else:
+        # This is impossible.
+        return 'Great scott! No time passed!'
```
Comment (ebolyen) [inline qiime/core/util.py]: `'0 %s' % attrs[-1]`

File: qiime/core/tests/test_util.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import tempfile
+import pathlib
+import collections
+import datetime
+import dateutil.relativedelta as relativedelta
+
+import qiime.core.util as util
+
+
+class TestDurationTime(unittest.TestCase):
```
Comment (ebolyen) [inline qiime/core/tests/test_util.py]: Test for negative duration

**Comment by ebolyen** [comment]
@jairideout ready for review/merge. I wasn't able to move the yaml representer on plugin to the provenance_capture as it resulted in a cyclical dependency. I'm sure there is some way to break it, but it doesn't seem worth it right now.

**Comment by jairideout** [comment]
Thanks @ebolyen, it is great to have true provenance support in QIIME 2! The future is now 🚀

**Pull request merged**